factorize.py

```diff
@@ -0,0 +1,78 @@
+import nimfa
+import pandas as pd
+from scipy.sparse import csr_matrix as csrm
+from data_prep_tools import positions_list
+
+def nmf_all_positions(df_with_points):
+    decomposition_results = [nmf_one_position(df_with_points, position) \
+                                for position in positions_list]
+    offense = pd.concat([results['offense'] for results in decomposition_results])
+    defense = pd.concat([results['defense'] for results in decomposition_results])
+    return offense, defense
+
+def nmf_one_position(df_with_points, position):
+    position_df = df_with_points.query('position == @position')
+    sparse_position_matrix, player_ids, opponents = sparsify(position_df)
+    offense_skill, defense_skill = decompose(sparse_position_matrix)
+    offense_skill_df = pd.DataFrame({   'player_id': player_ids,
+                                        'off_factorized_skill': offense_skill
+                                    })
+    defense_skill_df = pd.DataFrame({   'opponent': opponents,
+                                        'position': position,
+                                        'def_factorized_skill': defense_skill
+                                    })
+    return {'offense': offense_skill_df, 
+            'defense': defense_skill_df}
+
+def sparsify(input_df, clip_positive=True):
+    '''
+    Input:  DataFrame containing player_ids, opponent and fanduel points to be sparsified
+            clip_positive specifies whether to clip all values at 0 (typically for nmf)
+    Output: Sparse matrix (player count in position x opponents) of fanduel points scored,
+            list of player_ids for the sparse matrix, list of opponents for the sparse matrix
+    '''
+    if clip_positive:
+        input_df.fanduel_points = input_df.fanduel_points.clip(0)
```

Comment:
That ternary statement wasn't meant to operate element-wise, it's just storing a series in fanduel_points, a series which isn't a slice of the df.

I tried the test code that you wrote, I'm not sure what you we're trying to write, but it doesn't execute because a doesn't have a column 'bbb'.

I think that the origin of the issue comes form the fact that input_df comes from a query...I didn't think this would cause problems. But apparently it does, try:
a=pd.DataFrame({'aaa': range(10)})
a2 = a.query('aaa > 4')
a2.eval('bbb = aaa * 2')